src/controllers/discordController.ts

```diff
@@ -0,0 +1,176 @@
+import ClientOAuth2 from 'client-oauth2';
+import {Request, Response} from 'express';
+import axios, { AxiosRequestConfig, AxiosPromise } from 'axios';
+import {User, Role} from '../entity/User';
+import {getManager} from 'typeorm';
+import * as auth from '../middleware/authMiddleware';
+import { v4 as uuidv4 } from 'uuid';
+import {validate} from 'class-validator';
+import * as crypto from 'crypto';
+import {State} from '../entity/State';
+
+const discordAuth = new ClientOAuth2({
+    clientId: process.env.discordClientId,
+    clientSecret: process.env.discordClientSecret,
+    accessTokenUri: 'https://discord.com/api/oauth2/token',
+    authorizationUri: 'https://discord.com/api/oauth2/authorize',
+    scopes: ['identify', 'email']
+});
+
+
+export const authDiscord = async (req: Request, res: Response): Promise<void> => {
+    const state = crypto.randomBytes(16).toString('hex');
+
+    const stateRepo = getManager().getRepository(State);
+    await stateRepo.save(stateRepo.create({state: state, type: 'discord'}));
+    res.redirect(discordAuth.code.getUri({redirectUri: 'http://localhost:8080/api/auth/discord/callback/signin', state: state}));
+}
+
+export const linkDiscord = async (req: Request, res: Response): Promise<void> => {
+    const token = req.header('Authorization')?.split(' ')[1];
+    if(!token) {
+        res.sendStatus(401); // User was not properly authenticated...
+        return;
+    }
+
+    const uuid = auth.getUserFromToken(token);
+    if(!uuid) {
+        res.sendStatus(401); // User was not properly authenticated...
+        return;
+    }
+
+    const state = crypto.randomBytes(16).toString('hex') + '/' + uuid
+
+    const stateRepo = getManager().getRepository(State);
+    await stateRepo.save(stateRepo.create({state: state, type: 'discord'}));
+    res.redirect(discordAuth.code.getUri({redirectUri: 'http://localhost:8080/api/auth/discord/callback/link', state: state}));
+}
+
+export const discordAuthCallback = async (req: Request, res: Response): Promise<void> => {
+    // Before we do anything, check that the state is valid
+    const stateRepo = getManager().getRepository(State);
+    const state = req.query.state as string | undefined;
```

Comment:
Is `|` the correct operator here?

Reply:
It was being used as a type union since the query might not have a state param. I've changed it so that it explicitly checks that the state exists first.